File: Step 07 Recursion/Step 7.1 Get a Strong Hold/Count Good numbers/Count Good numbers.cpp
class Solution {
public:
    int countGoodNumbers(long long n) {
        const int MOD = 1e9 + 7;
        long long evenIndices = (n + 1) / 2;
        long long oddIndices = n / 2;
        long long result1 = myPow(5, evenIndices, MOD);
        long long result2 = myPow(4, oddIndices, MOD);
        return (int)((result1 * result2) % MOD);
    }

private:
    long long myPow(long long x, long long n, int MOD) {
        if (n == 0 || x == 1) return 1;
        if (x == 0) return 0;
        x = x % MOD;  // Handle cases where x is greater than MOD
        long long halfPow = myPow((x * x) % MOD, n / 2, MOD);
        return (n % 2 == 0) ? halfPow : (x * halfPow) % MOD;
    }
};

File: Step 07 Recursion/Step 7.1 Get a Strong Hold/Count Good numbers/Count Good numbers.java
class Solution {
    private static final int MOD = 1000000007;

    public int countGoodNumbers(long n) {
        long evenIndices = (n + 1) / 2;
        long oddIndices = n / 2;
        return (int) ((myPow(5, evenIndices) * myPow(4, oddIndices)) % MOD);
    }

    private long myPow(long x, long n) {
        if (n == 0) return 1;
        long result = 1;
        x = x % MOD;
        while (n > 0) {
            if (n % 2 == 1) {
                result = (result * x) % MOD;
            }
            x = (x * x) % MOD;
            n /= 2;
        }
        return result;
    }
}
